lib/src/message_window.dart

```diff
@@ -12,32 +14,76 @@ final _emptyData = Uint8List(0);
 
 typedef _ServerMessageFn = ServerMessage Function(Uint8List data);
 
-Map<int, _ServerMessageFn> _messageTypeMap = {
-  49: (d) => ParseCompleteMessage(),
-  50: (d) => BindCompleteMessage(),
-  65: NotificationResponseMessage.new,
-  67: CommandCompleteMessage.new,
-  68: DataRowMessage.new,
-  69: ErrorResponseMessage.new,
-  75: BackendKeyMessage.new,
-  82: AuthenticationMessage.new,
-  83: ParameterStatusMessage.new,
-  84: RowDescriptionMessage.new,
-  87: CopyBothResponseMessage.new,
-  90: ReadyForQueryMessage.new,
-  100: CopyDataMessage.new,
-  110: (d) => NoDataMessage(),
-  116: ParameterDescriptionMessage.new,
-  $3: (d) => CloseCompleteMessage(),
-};
+// Map<int, _ServerMessageFn> _messageTypeMap = {
+//   49: (d) => ParseCompleteMessage(),
+//   50: (d) => BindCompleteMessage(),
+//   65: NotificationResponseMessage.new,
+//   67: CommandCompleteMessage.new,
+//   68: DataRowMessage.new,
+//   69: ErrorResponseMessage.new,
+//   75: BackendKeyMessage.new,
+//   82: AuthenticationMessage.new,
+//   83: ParameterStatusMessage.new,
+//   84: RowDescriptionMessage.new,
+//   87: CopyBothResponseMessage.new,
+//   90: ReadyForQueryMessage.new,
+//   100: CopyDataMessage.new,
+//   110: (d) => NoDataMessage(),
+//   116: ParameterDescriptionMessage.new,
+//   $3: (d) => CloseCompleteMessage(),
+// };
 
 class MessageFramer {
   final _reader = ByteDataReader();
   final messageQueue = Queue<ServerMessage>();
+  final Encoding encoding;
+
+  _ServerMessageFn? _messageTypeMap(int? msgType) {
```

Comment:
Not sure: is there any performance benefit to changing to `switch`? I'd rather change the `_ServerMessageFn` interface to also accept `Encoding`, unless there is evidence for improvement here.

Reply:
A switch statement will usually be higher performance. It creates a lookup table and goto reference and starts at that point.

```dart
final statusMap = <int, String>{
  0: 'Initializing',
  1: 'Idle',
  2: 'Running',
  3: 'Stop',
  4: 'Error',
  5: 'a',
  6: 'b',
  7: 'c',
  8: 'd',
  9: 'e',
  10: 'f',
  11: 'g',
  12: 'h',
  13: 'i',
  14: 'j',
  15: 'k',
  16: 'l',
};

 String getStatusByMap(int value) {
  return statusMap[value]!;
}

String getStatusBySwitch(int value) {
  switch (value) {
    case 0:
      return 'Initializing';
    case 1:
      return 'Idle';
    case 2:
      return 'Running';
    case 3:
      return 'Stop';
    case 4:
      return 'Error';
    case 5:
      return 'a';
    case 6:
      return 'b';
    case 7:
      return 'c';
    case 8:
      return 'd';
    case 9:
      return 'e';
    case 10:
      return 'f';
    case 11:
      return 'g';
    case 12:
      return 'h';
    case 13:
      return 'i';
    case 14:
      return 'j';
    case 15:
      return 'k';
    case 16:
      return 'l';
    default:
      return 'Undefined';
  }
}

String getStatusByIfElse(int value) {
  if (value == 0) {
    return 'Initializing';
  } else if (value == 1) {
    return 'Idle';
  } else if (value == 2) {
    return 'Running';
  } else if (value == 3) {
    return 'Stop';
  } else if (value == 4) {
    return 'Error';
  } else if (value == 5) {
    return 'a';
  } else if (value == 6) {
    return 'b';
  } else if (value == 7) {
    return 'c';
  } else if (value == 8) {
    return 'd';
  } else if (value == 9) {
    return 'e';
  } else if (value == 10) {
    return 'f';
  } else if (value == 11) {
    return 'g';
  } else if (value == 12) {
    return 'h';
  } else if (value == 13) {
    return 'i';
  } else if (value == 14) {
    return 'j';
  } else if (value == 15) {
    return 'k';
  } else if (value == 16) {
    return 'l';
  } else {
    return 'Undefined';
  }
}

void measure(void Function(int value) cb, String name) {
  final stopwatch = Stopwatch()..start();
  for (var i = 0; i < 1000000000; i++) {
    final val = i % 5;
    cb(val);
  }
  print('$name ${stopwatch.elapsed}');
}

void main(List<String> args) {
  measure(getStatusByMap, 'Map');
  measure(getStatusByIfElse, 'If-Else');
  measure(getStatusBySwitch, 'Switch');
  
// Result:
// Map 0:00:12.680369
// If-Else 0:00:05.666660
// Switch 0:00:06.000052
}

```